Implement `whence=1` for the `utils.PackedObjectReader.seek`

sphuber opened the pull request:
Fixes #67 

This mode of seek is often used, for example by `numpy.load` and not
supporting it would prohibit a `Container` based repository with packed
objects from being used with it. Here we add support for `whence=1`
which simply means that the `target` is with respect to the current
byte position as reported by `tell` instead of with `0`.

giovannipizzi commented:
Looks good!

As we discussed, we also then need to implement this for the `utils.StreamDecompresser`.

Also, probably ideal to also have tests to test the minimal seek functionality at the container level, testing loose, packed uncompressed and packed compressed objects (with e.g. `pytest.mark.parametrize('compress', [True, False])`). Thanks!

PS1: Do you think we should also implement `whence = 2`?
PS2: While you are at it, I realised that in the `StreamDecompresser` we don't need to seek to zero every time. If we are seeking to a position that is > than the current one, we need need to read (and discard) `target - self.tell()`, right? In this way it's at least (relatively) efficient to seek forward, even if compressed. Maybe you can give a shot at this (or if you don't have the time, could you open an issue?)

sphuber commented:
Working on it. Will push an update later

sphuber commented:
@giovannipizzi I added the same functionality for the `StreamDecompresser` and added a parametrized test on a stream obtained through `Container.get_object_stream`. For the latter, I also wanted to add some tests that exception is raised when we seek beyond the limits, but got some weird behavior. I am not sure how the `io` classes are supposed to behave. For example:
```
In [1]: import io                                                                                                                                                                                                

In [2]: stream = io.StringIO('test')                                                                                                                                                                             

In [3]: stream.seek(-1)                                                                                                                                                                                          
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-6b01d185eb3f> in <module>
----> 1 stream.seek(-1)

ValueError: Negative seek position -1

In [4]: stream.seek(3)                                                                                                                                                                                           
Out[4]: 3

In [5]: stream.seek(-1, 1)                                                                                                                                                                                       
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-5-57cab1c18538> in <module>
----> 1 stream.seek(-1, 1)

OSError: Can't do nonzero cur-relative seeks
```

As you can see seeking negatively for `whence=0` raises a `ValueError` as expected. However, I am also not allowed to use negative offsets when using `whence=1` which raises an `OSError` . When opening a file through `open`, I get a different exception:
```
In [9]: with open('codecov.yml', 'r') as handle: 
   ...:     handle.seek(2) 
   ...:     handle.seek(-1, 1) 
   ...:      
   ...:                                                                                                                                                                                                          
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-9-4844d5b8816f> in <module>
      1 with open('codecov.yml', 'r') as handle:
      2     handle.seek(2)
----> 3     handle.seek(-1, 1)
      4 
      5 

UnsupportedOperation: can't do nonzero cur-relative seeks
```
Aside from the different exception (which is a subclass of `OSError`), I am confused that the negative offsets are not allowed for `whence=1`. [The documentation of `seek`](https://docs.python.org/3/library/io.html#io.IOBase.seek) clearly says that offsets *can* be negative for `whence=1`.

giovannipizzi commented:
You can seek back, but only for bytes objects, not for strings (because going backward reliably in strings it's a mess and maybe not even possible in general, I think).

With `io.BytesIO` and opening with `b` in the mode works instead:

```
In [1]: import io

In [2]: a = io.BytesIO(b'12345')

In [3]: a.seek(3)
Out[3]: 3

In [4]: a.read(1)
Out[4]: b'4'

In [5]: a.seek(-2,1)
Out[5]: 2

In [6]: a.read()
Out[6]: b'345'
```

```
In [8]: with open ('index.html', 'rb') as handle:
   ...:     handle.seek(5)
   ...:     print(handle.seek(-2, 1))
   ...:     
3
```

Do you really need `StringIO`? Maybe you meant to use `BytesIO`?

giovannipizzi commented:
BTW, please rebase on `develop` because I did a number of important changes that might collide maybe with yours, and also most importantly there was an important bug (fixed in #74) with writing directly to compressed packs that might be important to have correct in your PR

sphuber commented:
> Do you really need `StringIO`?

No, I was just testing what the interface is of the basic streams in the built-in library with respect to seeking. Anyway, I indeed forgot that seeking (bytes) of course becomes messy when it is not on bytes with multi-byte characters and all. So you can consider my confusion as solved :)

sphuber commented:
> BTW, please rebase on `develop` because I did a number of important changes that might collide maybe with yours, and also most importantly there was an important bug (fixed in #74) with writing directly to compressed packs that might be important to have correct in your PR

Done
